### PR TITLE
fix: writing 0 to accumulator when you mean to write 32 bit value

### PR DESCRIPTION
**Pull request by cbrisebois**
fix: comment
refactor: removed magic 32 numbers and replaced with constant Accumulator size.
feat: new tests for catching 32 bit accumulator edge cases

### CODE REVIEW

**Review by eriksvedang — APPROVED** [review]
This code is a little hard to critique without spending a lot of time understanding the exact details, but from afar it looks great. Removing magic values and adding tests is awesome!

**Comment by cbrisebois** [comment]
ok so glad you commented because it gave me a chance to think about it
more.  The MaskFromCount method is a clever, clever way to make a bit mask
from a count of bits.  So if you want a mask like 0x0007FFF you pass in 15
and a nice 15 bit mask is created because it shifts 1 to the 15th position
and then subtracts 1 which does the clever trick.  In the case of a count
of 32, you expect a mask of 0xFFFFFFFF but the C# version of this returns a
mask of 0.  (I'll explain in a minute).  The expected behaviour of this
method is shift 1 << 32 which, in a 32bit register, would be 0x00000000
then you subtract 1 to get 0xFFFFFFFF.  So now that I'm looking at it and
I've read the Golang spec more it should actually be ok and not need the 32
bit check there.  Golang backs all integer numbers on a 64 bit register so
unless we have a 64 bit accumulator this wont happen, so i'll back out that
change.

The worse problem is the C# version.  The expected behaviour is this:

(1<<32) = 0x00000000

what we get is:

(1<<32) = 0x00000001

So when I subtract 1 the mask is 0x00000000, not 0xFFFFFFFF so the whole
stream breaks down.  So the << in C# is behaving like a mod, not a pure
register shift.  The MS .net spec seems to indicate it should not do this,
but I guess it depends on the version and compiler.

TL;DR, i'm going to take the condition check out of MaskFromCount in this
version.  Thanks for reading.

**Comment by eriksvedang** [comment]
@cbrisebois Glad to be of some use :D 

And thanks for the longer explanation – now I actually understand what the code is trying to do. I'll be happy to re-review this once you've made the changes!

**Comment by eriksvedang** [comment]
Just for fun I tried this in C, using clang and got this:

```c
int x = 1 << 32;
```

```bash
shift.c:4:15: warning: shift count >= width of type [-Wshift-count-overflow]
    int x = 1 << 32;
              ^  ~~
1 warning generated.
```

```bash
$ ./a.out
349ef60
```

So it seems like it's undefined behaviour in C?

**Comment by eriksvedang** [comment]
Not that it matters much how it works in C, but just thought it was interesting to compare...

**Comment by cbrisebois** [comment]
yeah, so you can't really do it in C# or go either, you get overflow
warnings and have to compile in C# with "unchecked" but at runtime.  My
recollection of C is that it would shift into oblivion and you'd have a
result of 0, but it has been some time since I used C.  After our
discussion I'm going to just leave in the condition since it makes the
operations really clear and doesn't rely on language / compiler behaviour.

**Comment by eriksvedang** [comment]
@cbrisebois Yeah, sounds like the best idea.
